service/controller/resource/nodes/resource.go

```diff
@@ -0,0 +1,145 @@
+package nodes
+
+import (
+	"context"
+
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/encrypter"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+
+	// "github.com/giantswarm/azure-operator/v4/service"
+	"github.com/giantswarm/azure-operator/v4/service/controller/debugger"
+	"github.com/giantswarm/azure-operator/v4/service/controller/internal/state"
+	"github.com/giantswarm/azure-operator/v4/service/controller/internal/vmsscheck"
+	"github.com/giantswarm/azure-operator/v4/service/controller/setting"
+)
+
+type Config struct {
+	Debugger  *debugger.Debugger
+	G8sClient versioned.Interface
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+
+	Azure            setting.Azure
+	InstanceWatchdog vmsscheck.InstanceWatchdog
+	Name             string
+}
+
+type Resource struct {
+	debugger     *debugger.Debugger
+	g8sClient    versioned.Interface
+	k8sClient    kubernetes.Interface
+	logger       micrologger.Logger
+	stateMachine state.Machine
+
+	azure            setting.Azure
+	instanceWatchdog vmsscheck.InstanceWatchdog
+	name             string
+}
+
+func New(config Config) (*Resource, error) {
+	if config.Debugger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Debugger must not be empty", config)
+	}
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
+	}
+	if config.InstanceWatchdog == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.InstanceWatchdog must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if err := config.Azure.Validate(); err != nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Azure.%s", config, err)
+	}
+
+	if len(config.Name) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Name must not be empty", config)
+	}
+
+	r := &Resource{
+		debugger:  config.Debugger,
+		g8sClient: config.G8sClient,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		azure:            config.Azure,
+		instanceWatchdog: config.InstanceWatchdog,
+		name:             config.Name,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Azure() setting.Azure {
+	return r.azure
+}
+
+func (r *Resource) Debugger() *debugger.Debugger {
+	return r.debugger
+}
+
+func (r *Resource) G8sClient() versioned.Interface {
+	return r.g8sClient
+}
+
+func (r *Resource) InstanceWatchdog() vmsscheck.InstanceWatchdog {
+	return r.instanceWatchdog
+}
+
+func (r *Resource) Logger() micrologger.Logger {
```

Comment:
Do you have some other use for these getters than just accessing the base struct fields? I don't think they are needed for that and especially the logging lines earlier in this PR look a bit unusual.

See https://play.golang.org/p/apUb8sVz26T

Could we live without these and just directly use the fields since they are embedded in the "derived" struct?

Reply:
It's probably more due to my brain's wiring from other programming languages than to some strict idea. :smile: 

The reasoning here was to leave fields in base struct private and expose them via getter functions, but since this is the case of a straightforward embedding, without some business logic around getting those fields, making fields public in the base struct and exposing them is simpler and clearer. Changing it now to exposed `.Logger` field in the base struct.

Reply:
Updated to use exposed `.Logger` field (did the same for other fields / getters).